src/utils/gcsApi.ts

```diff
@@ -0,0 +1,301 @@
+      logger.debug(`Using configuration located at ${filepathVar.name}`);
+      if (!fs.existsSync(gcsCredsPath)) {
+        reportError(`File does not exist: ${gcsCredsPath}`);
+      }
+      configRaw = fs.readFileSync(gcsCredsPath).toString();
+    } else {
+      const errorMsg =
+        `GCS credentials not found! Please provide the path to the ` +
+        `configuration via environment variable ${filepathVar.name}, or ` +
+        `specify the entire configuration in ${jsonVar.name}.`;
+      reportError(errorMsg);
+    }
+
+    let parsedCofig;
+    try {
+      parsedCofig = JSON.parse(configRaw);
+    } catch (err) {
+      reportError('Error parsing JSON credentials');
+    }
+
+    const { project_id, client_email, private_key } = parsedCofig;
+
+    for (const field of [project_id, client_email, private_key]) {
+      if (!field) {
+        reportError(`GCS credentials missing ${field}!`);
+      }
+    }
+
+    return { project_id, client_email, private_key };
+  };
+
+  /**
+   * Detect the content-type based on regular expressions defined in
+   * CONTENT_TYPES_EXT.
+   *
+   * The underlying GCS package usually detects content-type itself, but it's
+   * not always correct.
+   *
+   * @param artifactName Name of the artifact to check
+   * @returns A content-type string, or undefined if the artifact name doesn't
+   * have a known extension
+   */
+  private detectContentType(artifactName: string): string | undefined {
+    for (const entry of CONTENT_TYPES_EXT) {
+      const [regex, contentType] = entry;
+      if (artifactName.match(regex)) {
+        return contentType;
+      }
+    }
+    return undefined;
+  }
+
+  /**
+   * Uploads the artifact at the given local path to the path specified in the
+   * given config object
+   *
+   * @param localFilePath Location of the file to be uploaded
+   * @param uploadConfig Configuration for the upload including destination path
+   * and metadata
+   */
+  private async uploadArtifactFromPath(
+    localFilePath: string,
+    // require these three properties out of the GCSUploadOptions interface
+    uploadConfig: Pick<
+      Required<GCSUploadOptions>,
+      'destination' | 'metadata' | 'gzip'
+    >
+  ): Promise<void> {
+    const destinationFilePath = uploadConfig.destination as string;
+    if (!destinationFilePath) {
+      return Promise.reject(
+        new Error(
+          `Can't upload file at ${localFilePath} - no destination path specified!`
+        )
+      );
+    }
+
+    const destinationPath = path.dirname(destinationFilePath);
+    const filename = path.basename(localFilePath);
+    // const fileUploadConfig: GCSUploadOptions = { ...uploadConfig };
+    // const contentType = this.detectContentType(filename);
+    // if (contentType) {
+    //   fileUploadConfig.contentType = contentType;
+    // }
+    const fileUploadConfig: GCSUploadOptions = {
+      ...uploadConfig,
+      contentType: this.detectContentType(filename),
```

Comment:
Don't remember the exact behavior, but there's a chance that `contentType: undefined` might cause problems (hence the previous, commented-out implementation).

Reply:
Yes, thanks for the reminder. The commented out part in fact is still there specifically so I remember to test out whether or not that's the case (and so I can easily switch it back if the current implementation does cause problems), but I should have put a `TODO` also.

Will resolve this before finalizing the PR.

Reply:
You were right - it wasn't what I wanted, but I finally remembered the syntax I was looking for:

```
    const contentType = this.detectContentType(filename);
    const fileUploadConfig: GCSUploadOptions = {
      ...uploadConfig,
      contentType: this.detectContentType(filename),
      ...(contentType && { contentType }),
    };
```

Tested and working both when `contentType` is defined and when it's not.